server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -2584,6 +2585,52 @@ protected void runInContext() {
         }
     }
 
+    private void verifyVmLimits(UserVmVO vmInstance, Map<String,String> details) {
+        if (VirtualMachineManager.ResoureCountRunningVMsonly.value()) {
+            return;
+        }
+
+        long newCpu = NumberUtils.toLong(details.get(VmDetailConstants.CPU_NUMBER));
+        long newMemory = NumberUtils.toLong(details.get(VmDetailConstants.MEMORY));
+        ServiceOfferingVO currentServiceOffering = _serviceOfferingDao.findByIdIncludingRemoved(vmInstance.getId(), vmInstance.getServiceOfferingId());
+        ServiceOfferingVO svcOffering = _serviceOfferingDao.findById(vmInstance.getServiceOfferingId());
+        boolean isDynamic = currentServiceOffering.isDynamic();
+        if (isDynamic) {
+            Map<String, String> customParameters = new HashMap<>();
+            customParameters.put(VmDetailConstants.CPU_NUMBER, String.valueOf(newCpu));
+            customParameters.put(VmDetailConstants.MEMORY, String.valueOf(newMemory));
+            validateCustomParameters(svcOffering, customParameters);
+        }
+        long currentCpu = currentServiceOffering.getCpu();
+        long currentMemory = currentServiceOffering.getRamSize();
+
+        Account owner = _accountDao.findById(vmInstance.getAccountId());
+        if (owner == null) {
+            throw new InvalidParameterValueException("The owner of " + vmInstance + " does not exist: " + vmInstance.getAccountId());
+        }
+        try {
+            if (newCpu > currentCpu) {
+                _resourceLimitMgr.checkResourceLimit(owner, ResourceType.cpu, newCpu - currentCpu);
+            }
+            if (newMemory > currentMemory) {
+                _resourceLimitMgr.checkResourceLimit(owner, ResourceType.memory, newMemory - currentMemory);
+            }
+        } catch (ResourceAllocationException e) {
+            throw new CloudRuntimeException("Failed to update VM settings", e);
```

Comment:
@Pearl1594 use InvalidParameterValueException instead of CloudRuntimeException ?
this can be improved 

![Screenshot from 2021-09-17 12-00-28](https://user-images.githubusercontent.com/57355700/133764958-280760c4-bc33-4fc4-9f00-e4214b1363ef.png)

Reply:
Update @weizhouapache 
Looks like this now:
If we exceed the threshold set for a domain:
![image](https://user-images.githubusercontent.com/10495417/133767651-71cba77a-a11d-42e4-89ad-20be187d4a44.png)

If we exceed the limit enforced by custom offering:
![image](https://user-images.githubusercontent.com/10495417/133767723-9236ecf9-7ed3-4939-8406-898d747dbc5b.png)